filters/scheduler/fifo.go

```diff
@@ -139,14 +164,26 @@ func (f *fifoFilter) Request(ctx filters.FilterContext) {
 // Response will decrease the number of inflight requests to release
 // the concurrency reservation for the request.
 func (f *fifoFilter) Response(ctx filters.FilterContext) {
-	pending, ok := ctx.StateBag()[fifoKey].([]func())
-	if !ok {
-		return
+	g := f.createResponse(ctx)
+	switch f.typ {
+	case fifo:
+		g()
+	case fifoWithBody:
+		ctx.StateBag()[filters.FifoWithBody] = g
```

Comment:
Its not clear how multiple fifoWithBody would chain and how they would interplay with regular fifo.
I think you'd need to have the same logic stored under different keys and then run chain of hooks in two places in proxy depending on the key at proxy.do and proxy.ServeHTTP:


```go
...
func (f *fifoFilter) createResponse(ctx filters.FilterContext, typ string) {
		pending, ok := ctx.StateBag()[typ].([]func())
		if !ok {
			return
		}
		last := len(pending) - 1
		if last < 0 {
			return
		}
		pending[last]()
		ctx.StateBag()[fifoKey] = pending[:last]
}

...
// existing in proxy.do
pendingFIFO, _ := stateBag[scheduler.FIFOKey].([]func())
for _, done := range pendingFIFO {
	done()
}
...
// in proxy.ServeHTTP
pendingFIFO, _ := stateBag[scheduler.FIFOWithBodyKey].([]func())
for _, done := range pendingFIFO {
	done()
}
```

Reply:
Yes that's a good point. I also thought about that I need to do something like that.
What do you like more to support the copystream part? this one or the other?

Reply:
I think we can remove existing pendingFIFO from proxy.go see #2680